tests/test-prune.sh

```diff
@@ -0,0 +1,292 @@
+    # Wrap this to avoid set -x showing the ls commands
+    {
+        NUM_FILE=$(ls -d $1/objects/*/*.filez | wc -l)
+        NUM_DIRTREE=$(ls -d $1/objects/*/*.dirtree | wc -l)
+        NUM_COMMIT=$(ls -d $1/objects/*/*.commit | wc -l)
+        NUM_DIRMETA=$(ls -d $1/objects/*/*.dirmeta | wc -l)
+        NUM_COMMITMETA2=0
+        if compgen -G "$1/objects/*/*.commitmeta2" > /dev/null; then
+            NUM_COMMITMETA2=$(ls -d $1/objects/*/*.commitmeta2 | wc -l)
+        fi
+        NUM_OBJECT=$(ls -d $1/objects/*/* | wc -l)
+        echo OBJCOUNT $1: $NUM_FILE files + $NUM_DIRTREE dirtree + $NUM_COMMIT commit + $NUM_DIRMETA dirmeta + $NUM_COMMITMETA2 commitmeta2 == $NUM_OBJECT objects
+    } 2> /dev/null
+}
+
+
+ostree --repo=orig-repo --mode=archive init
+
+create_commit orig-repo app1 3 #   8f 7d 3c 2m == 20
+create_commit orig-repo app2 4 #+ 10f 9d 4c 0m == 23
+create_commit orig-repo app3 3 #+  8f 7d 3c 0m == 18
+#####################   26 23 10  2     61
+
+count_objects orig-repo
+assert_streq $NUM_FILE 26
+assert_streq $NUM_DIRTREE 23
+assert_streq $NUM_COMMIT 10
+assert_streq $NUM_DIRMETA 2
+assert_streq $NUM_OBJECT 61
+assert_streq $NUM_COMMITMETA2 0
+
+#################### Try a no-op prune, should do nothing but create commitmeta2 objects
+
+cp -ra orig-repo repo # Work on a copy
+
+# Prune with full depth should change nothing
+$FLATPAK build-update-repo --no-update-summary --no-update-appstream --prune --prune-depth=-1 repo > prune.log
+cat prune.log
+
+assert_file_has_content prune.log "Total objects: 61"
+
+count_objects repo
+assert_streq $NUM_FILE 26
+assert_streq $NUM_DIRTREE 23
+assert_streq $NUM_COMMIT 10
+assert_streq $NUM_DIRMETA 2
+assert_streq $NUM_COMMITMETA2 $NUM_COMMIT
+assert_streq $NUM_OBJECT 71
+
+cp -ra repo incremental-repo # Make a copy of the orig repo with the commitmeta2 objects
+
+# Try again with the commitmeta existing
+
+# Prune with full depth should change nothing
+$FLATPAK build-update-repo --no-update-summary --no-update-appstream --prune --prune-depth=-1 repo > prune.log
+cat prune.log
+
+assert_file_has_content prune.log "Total objects: 61"
+
+count_objects repo
+assert_streq $NUM_FILE 26
+assert_streq $NUM_DIRTREE 23
+assert_streq $NUM_COMMIT 10
+assert_streq $NUM_DIRMETA 2
+assert_streq $NUM_COMMITMETA2 $NUM_COMMIT
+assert_streq $NUM_OBJECT 71
+
+ok "no-op prune"
+
+############# Try various depth prunes, with and without .commitmeta2
+
+rm -rf repo
+cp -ra orig-repo repo # Work on a copy
+
+# depth = 2 will only remove one commit from app2
+
+$FLATPAK build-update-repo --no-update-summary --no-update-appstream --prune --prune-depth=2 repo > prune.log
+cat prune.log
+
+assert_file_has_content prune.log "Total objects: 61"
+assert_file_has_content prune.log "Deleted 5 objects,"
+
+count_objects repo
+assert_streq $NUM_FILE 24
+assert_streq $NUM_DIRTREE 21
+assert_streq $NUM_COMMIT 9
+assert_streq $NUM_DIRMETA 2
+assert_streq $NUM_COMMITMETA2 $NUM_COMMIT
+assert_streq $NUM_OBJECT 65
```

Comment:
Maybe add a comment to explain that it's 65 even though 71 - 5 = 66 because of the commitmeta2 object not being counted

Reply:
I don't know how to phrase any of the computations of the numbers understandably here, but I will add a mention of the commitmeta in the create_commit() function.